index.js

```diff
@@ -110,39 +129,77 @@ async function testRetrieval({
       )
       await new Promise((resolve) => setTimeout(resolve, retryDelayMs))
       return testRetrieval({
+        pdpVerifier,
+        pandoraService,
         clientAddress,
         CDN_HOSTNAME,
         rootCid,
-        isSupportedSP,
         setId,
         rootId,
-        ownerUrl,
         retryOn404: false,
       })
     }
 
-    if (isSupportedSP) {
-      console.error(
-        'ALERT Cannot retrieve ProofSet %s Root %s SP %s via %s: %s %s',
-        String(setId),
-        String(rootId),
-        URL.parse(ownerUrl)?.hostname ?? ownerUrl,
-        url,
-        res.status,
-        reason,
-      )
-    }
+    const proofSetIdHeaderValue = res.headers.get('x-proof-set-id')
+    const pieceRetrievalUrl = await maybeGetResolvedProofSetRetrievalUrl({
+      pdpVerifier,
+      pandoraService,
+      proofSetIdHeaderValue,
+    })
+
+    console.error(
+      'ALERT Cannot retrieve ProofSet %s Root %s (resolved as ProofSet %s from SP %s) via %s: %s %s',
+      String(setId),
+      String(rootId),
+      proofSetIdHeaderValue ?? '(not reported)',
+      pieceRetrievalUrl
+        ? (URL.parse(pieceRetrievalUrl)?.hostname ?? pieceRetrievalUrl)
+        : '(unknown)',
+      url,
+      res.status,
+      reason,
+    )
   } else if (res.body) {
     const reader = res.body.getReader()
     while (true) {
       const { done } = await reader.read()
       if (done) break
     }
+  }
+}
+
+/**
+ * @param {object} args
+ * @param {PdpVerifier} args.pdpVerifier
+ * @param {PandoraService} args.pandoraService
+ * @param {string | null} args.proofSetIdHeaderValue
+ * @returns {Promise<string | undefined>} The piece retrieval URL
+ */
+async function maybeGetResolvedProofSetRetrievalUrl({
+  pdpVerifier,
+  pandoraService,
+  proofSetIdHeaderValue,
+}) {
+  if (proofSetIdHeaderValue === null || proofSetIdHeaderValue === '') {
+    return undefined
+  }
 
-    // NOTE: Even if the SP (ProofSet owner) is not supported, the retrieval can still
-    // succeed in case that somebody else stored the same file with a participating SP.
-    // For that reason, we are not alerting when this happens.
+  let proofSetId
+  try {
+    proofSetId = BigInt(proofSetIdHeaderValue)
+  } catch (err) {
+    console.warn(
+      'FilCDN reported invalid ProofSetID %j: %s',
+      proofSetIdHeaderValue,
+      err,
+    )
+    return undefined
   }
+
+  const [proofSetOwner] = await pdpVerifier.getProofSetOwner(proofSetId)
```

Comment:
I wonder if we should also handle errors in case `pdpVerifier` or `pandoraService` are not reachable and return `undefined` as we're doing for parsing the `proofSetId` to `BigInt`?